### PR TITLE
ARROW-6347: [GLib] Add garrow_array_diff_unified()

### PR DESCRIPTION
**Pull request by shiro615**

### CODE REVIEW

**Review by bkietz — CHANGES_REQUESTED** [review]
Out of curiosity, what is your use case for array diff? On the C++ side it's used for reporting from array comparison, so maybe you just need to set [`EqualOptions::diff_sink`](https://github.com/apache/arrow/blob/master/cpp/src/arrow/compare.h#L62-L64)?

If you really need to do more with diffs than format them, I'd recommend also wrapping `VisitEditScript()` and testing against that instead of reimplementing handling of the edit script format (`struct<run_length:int64, insert:bool>`)

**Comment by shiro615** [comment]
Thank you for your comments.
I was only looking at the low-level API.  I think we should set `EqualOptions :: diff_sink` as you say.
I'll fix to output formatted diffs.

**Comment by kszucs** [comment]
@ursabot build

**Comment by shiro615** [comment]
@kou
I've set `EqualOptions :: diff_sink` in `arrow_array_diff()` to treat as a convenience function.
Could you review this?

File: c_glib/arrow-glib/basic-array.cpp [hunk 1]
```diff
@@ -544,6 +544,30 @@ garrow_array_view(GArrowArray *array,
   }
 }
 
+/**
+ * garrow_array_diff:
```
Comment (kou) [inline c_glib/arrow-glib/basic-array.cpp]: Could you rename diff to `garrow_array_diff_unified`?
We'll provide low-level `garrow_array_diff` API later.

File: c_glib/arrow-glib/basic-array.cpp [hunk 2]
```diff
@@ -544,6 +544,30 @@ garrow_array_view(GArrowArray *array,
...
+gchar *
+garrow_array_diff(GArrowArray *array, GArrowArray *other_array)
+{
+  const auto arrow_array = garrow_array_get_raw(array);
+  const auto arrow_other_array = garrow_array_get_raw(other_array);
+  std::stringstream diff;
+  arrow_array->Equals(arrow_other_array,
+                      arrow::EqualOptions().diff_sink(&diff));
+  auto string = diff.str();
+  return g_strndup(string.data(), string.size());
```
Comment (kou) [inline c_glib/arrow-glib/basic-array.cpp]: Could you return `NULL` when no difference?

```cpp
if (string.empty()) {
  return NULL;
} else {
  return g_strndup(string.data(), string.size());
}
```

File: c_glib/arrow-glib/basic-array.cpp [hunk 3]
```diff
@@ -544,6 +544,30 @@ garrow_array_view(GArrowArray *array,
   }
 }
 
+/**
+ * garrow_array_diff:
+ * @array: A #GArrowArray.
+ * @other_array: A #GArrowArray to be compared.
+ *
+ * Returns: (transfer full): The string representation of
+ *   the difference between two arrays.
```
Comment (kou) [inline c_glib/arrow-glib/basic-array.cpp]: Could you add "as unified format"?

File: c_glib/test/test-array.rb [hunk 4]
```diff
@@ -131,4 +131,24 @@ def test_invalid
...
+      array = build_string_array(["Start", "Shutdown", "Reboot"])
+      other_array = build_string_array(["Start", "Shutdonw", "Reboot"])
+      assert_equal(<<-STRING.chomp, array.diff(other_array))
+
+@@ -1, +1 @@
+-"Shutdown"
++"Shutdonw"
+
+      STRING
+    end
```
Comment (kou) [inline c_glib/test/test-array.rb]: Could you also add different type case? For example, string array and int8 array.

**Review by kou — CHANGES_REQUESTED** [review]
Could you check my comments?

File: c_glib/arrow-glib/basic-array.cpp (hunk 1 above)
Comment (shiro615) [reply c_glib/arrow-glib/basic-array.cpp]: I've renamed it.

File: c_glib/arrow-glib/basic-array.cpp (hunk 2 above)
Comment (shiro615) [reply c_glib/arrow-glib/basic-array.cpp]: I've fixed it.

File: c_glib/arrow-glib/basic-array.cpp (hunk 3 above)
Comment (shiro615) [reply c_glib/arrow-glib/basic-array.cpp]: I've fixed it.

File: c_glib/test/test-array.rb (hunk 4 above)
Comment (shiro615) [reply c_glib/test/test-array.rb]: I've added test cases for int8 array.

File: c_glib/arrow-glib/basic-array.cpp [hunk 5]
```diff
@@ -544,6 +544,34 @@ garrow_array_view(GArrowArray *array,
...
 
+/**
+ * garrow_array_diff_unified:
+ * @array: A #GArrowArray.
+ * @other_array: A #GArrowArray to be compared.
+ *
+ * Returns: (nullable) (transfer full): The string representation of
+ *   the difference between two arrays as unified format.
+ *
+ *   It should be freed with g_free() when no longer needed.
```
Comment (kou) [inline c_glib/arrow-glib/basic-array.cpp]: Could you add a description for `NULL` case?

File: c_glib/test/test-array.rb (hunk 4 above)
Comment (kou) [reply c_glib/test/test-array.rb]: Ah, I want a test case for different type arrays case. For example, `string_array.diff_unified(int8_array)`.

Comment (shiro615) [reply c_glib/test/test-array.rb]: Oh...sorry.  I've also added a test case for different type arrays case.

File: c_glib/arrow-glib/basic-array.cpp (hunk 5 above)
Comment (shiro615) [reply c_glib/arrow-glib/basic-array.cpp]: I've added a description.

File: c_glib/test/test-array.rb [hunk 6]
```diff
@@ -131,4 +131,31 @@ def test_invalid
       assert_equal(message, error.message.lines.first.chomp)
     end
   end
+
+  sub_test_case("#diff_unified") do
+    def test_no_diff
+      array = build_string_array(["Start", "Shutdown", "Reboot"])
+      other_array = build_string_array(["Start", "Shutdown", "Reboot"])
+      assert_equal(nil, array.diff_unified(other_array))
```
Comment (kou) [inline c_glib/test/test-array.rb]: `assert_nil(array.diff_unified(other_array))` is better.

**Review by kou — APPROVED** [review]
+1
We can merge this once CI is green.

File: c_glib/test/test-array.rb (hunk 6 above)
Comment (shiro615) [reply c_glib/test/test-array.rb]: Thanks, fixed now.

**Pull request closed without merging**